ceph-disk: ability to use a different cluster name with dmcrypt

leseb opened the pull request:
Prior to this commit we were not able to configure an OSD using dmcrypt
on a cluster with a different name than 'ceph'. Adding the command line
option to --cluster <cluster name> to fix this.

Signed-off-by: Sébastien Han <seb@redhat.com>

alfredodeza commented on src/ceph-disk/ceph_disk/main.py:
```diff
@@ -2377,13 +2379,14 @@ def set_or_create_partition(self):
             self.partition = self.create_partition()
 
-    def create_key(self):
+    def create_key(self, cluster):
```
there is no need for requiring cluster here, you can rely on `self.args` since that will have the arguments passed in:

    self.args.cluster

alfredodeza commented on src/ceph-disk/ceph_disk/main.py:
```diff
@@ -2377,13 +2379,14 @@ def set_or_create_partition(self):
         command_check_call(
             [
                 'ceph',
+                '--cluster', cluster,
```
you would be able to just use that `self.args.cluster` in here, which already defaults to `ceph` if not explicitly defined

ktdreyer commented:
Would you please file a ticket at http://tracker.ceph.com for this, and add the URL of the ticket to the commit log, so we can track this and ensure it is backported to Jewel eventually?

leseb commented:
@ktdreyer sure, see: http://tracker.ceph.com/issues/17821

leseb commented:
I'll update the PR with a definitive fix today.

tchaikov commented:
@leseb could you put the line of `Fixes: http://tracker.ceph.com/issues/17821` into your commit message, right before your "SIgned-off-by" line, see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#tag-the-commit .

leseb commented:
@tchaikov just did, thanks!
I still have some work to do on that patch.

ktdreyer commented:
Please remove "DNM" from the title when this PR is ready for review and merging.

leseb commented:
@ErwanAliasr1 and I reworked the patch, ready for review now :)

ErwanAliasr1 commented:
WDYT about this patch series guys ?

ktdreyer commented:
Would you please rebase this onto the latest master to ensure that Jenkins is happy with it?

leseb commented:
@ktdreyer done

dachary commented:
this needs rebasing (and the test fails)

dachary commented on src/ceph-disk/ceph_disk/main.py:
```diff
@@ -4743,6 +4758,12 @@ def make_activate_parser(subparsers):
+        metavar='NAME',
+        default='ceph',
+        help='ceph cluster name (default: ceph)',
+    )
```
Activate does not have a --cluster argument because it can be determined with the find_cluster_by_uuid() function

leseb replied:
Alright, let me try again without this, I forgot the error that made us think we needed to do this.

leseb commented:
Ok testing is blocked by https://github.com/ceph/ceph/pull/12033/files waiting for the backport to re-validate.

leseb commented:
@dachary I just pushed a new version, just removed the commit that brings --cluster support to activate since this is not needed.

alfredodeza commented:
Jenkins test this please

alfredodeza commented:
@leseb looks like you have some valid test failures:

    E       AssertionError: Expected call: get_dmcrypt_key('UUID', '/etc/ceph/dmcrypt-keys', True)
    E       Actual call: get_dmcrypt_key('UUID', '/etc/ceph/dmcrypt-keys', True, 'ceph')

leseb commented:
@alfredodeza where can I get more details about these errors?

alfredodeza commented:
@leseb the logs expired, we need to try this again

alfredodeza commented:
jenkins test this please

leseb commented:
@alfredodeza not sure if my patch is responsible for this error, @dachary do you mind having a quick look into this please? Thanks!

alfredodeza commented:
@leseb these failures are related to the changes you've made. Specifically to the `get_dmcrypt_key()` function, which it has been altered to receive the extra argument of `'ceph'`.

That is why the test fails, because it was crafted to expect a strict set of arguments that didn't include the cluster name.

Since that is no longer the case, the test fails. You need to update the test to make sure that `'ceph'` is now expected.

The test failure occurs in: ../src/ceph-disk/tests/test_prepare.py line 319 

The actual failure is:

    E       AssertionError: Expected call: get_dmcrypt_key('UUID', '/etc/ceph/dmcrypt-keys', True)
    E       Actual call: get_dmcrypt_key('UUID', '/etc/ceph/dmcrypt-keys', True, 'ceph')

The assertion that needs to change is:

            assert m['_dmcrypt_map'].called
            m['get_dmcrypt_key'].assert_called_with(
               uuid, '/etc/ceph/dmcrypt-keys', True)

leseb commented:
@alfredodeza thanks for the detailed explanation, just pushed a change that should hopefully fix this issue.

leseb commented:
jenkins test this please

leseb commented:
@alfredodeza all green now :)
Can someone merge this? Btw @dachary I addressed your comment so if you're ok with the patch please approve the PR :)
Thanks!

dachary commented on src/ceph-disk/ceph_disk/main.py:
```diff
@@ -2443,7 +2451,7 @@ def populate(self):
         command_check_call(args)
         write_one_line(path, 'osd-uuid', self.args.osd_uuid)
-        self.create_key()
+        self.create_key(self.args.cluster)
```
the self.create_key method should use self.args.cluster: there is no need to pass it as an argument

leseb replied:
done

dachary commented:
Did you run a ceph-disk suite (teuthology --suite ceph-disk) successfully with this pull request ?

leseb commented:
@dachary I have not, can you help me with this?

dachary commented:
@leseb just push your branch to ceph-ci and I'll run it for you.

leseb commented:
@dachary like this? https://github.com/ceph/ceph-ci/pull/4

neurodrone commented:
It'd be nice to have this as a part of 10.2.6.

leseb commented:
@dachary can you run the CI with https://github.com/ceph/ceph-ci/pull/4?

dachary commented:
@leseb you need to push a branch on ceph-ci for teuthology to be run

leseb commented:
@dachary I don't have permission to push a branch to that repo, that's why I created a PR...

dachary commented:
@leseb you have permission to push, all ceph developers have permission and you're in the group, it must be something on your side that's not configured right.

leseb commented:
@dachary please see:

```
$ git remote -v |grep ceph-cii
ceph-cii        git@github.com:ceph/ceph-ci.git (push)
ceph-cii        git@github.com:ceph/ceph-ci.git (fetch)
$ git push ceph-cii dmcrypt-cluster-name
ERROR: Permission to ceph/ceph-ci.git denied to leseb.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Am I doing something wrong?

dachary commented:
I don't see what's wrong. Here are the permissions you have:

![seb](https://cloud.githubusercontent.com/assets/433594/22735835/d4ad1f82-edfb-11e6-912e-de5e3ca20aa2.png)

I guess someone with better knowledge of github subtle settings can help out :-)

leseb commented:
@dachary seems to be working now, branch pushed :)

dachary commented:
@leseb nice :-) What was the problem ? Do you have access to the sepia lab ?

leseb commented:
@dachary well looks like I wasn't part of Ceph team, then I got an email saying you added me and it was good :).

dachary commented:
> well looks like I wasn't part of Ceph team, then I got an email saying you added me and it was good :).

Oo amazing. I was looking to verify if you were listed and mistook the "add" box for the "search" box... happy ending :-) Welcome to the Ceph team, always nice to have new members ...

yuriw commented:
@leseb @dachary @GregMeno 
tests failed http://pulpito.front.sepia.ceph.com/yuriw-2017-02-15_21:45:11-ceph-disk-wip-yuri-testing2_2017_2_16-distro-basic-vps/

wip-yuri-testing2_2017_2_16

GregMeno commented:
http://pulpito.ceph.com/yuriw-2017-02-15_21:45:11-ceph-disk-wip-yuri-testing2_2017_2_16-distro-basic-vps/

GregMeno commented:
http://pulpito.ceph.com/yuriw-2017-02-15_21:45:11-ceph-disk-wip-yuri-testing2_2017_2_16-distro-basic-vps/

neurodrone commented:
Will this be backported to Jewel by any chance?

vumrao commented:
@neurodrone this is the tracker - http://tracker.ceph.com/issues/17821, it was marked backport to jewel and I have changed it to pending backport. It will get backported to jewel.

vasukulkarni commented:
@dachary it would have been nice to default the cluster to 'ceph' here, we got a patch for ceph-deploy to add the cluster option there, Also would be nice to run ceph-deploy suite for ceph-disk changes as well.

ganeshmaharaj commented:
@dachary @vasukulkarni Should have updated both threads. https://github.com/ceph/ceph/pull/13527 is the patch submitted to ceph-disk to add the default.

dachary commented:
@leseb I was wrong merging this pull request. I misread https://github.com/ceph/ceph/pull/11786#issuecomment-280190074 and thought the tests were successfull. Instead they **failed** and as a result ceph-disk is now broken in master. Could you please fix this asap ?

leseb commented:
@dachary I originally proposed the same approach as in https://github.com/ceph/ceph/pull/13527 so we should merge it.

dachary commented:
@leseb of course, entirely my fault, please accept my apologies

ganeshmaharaj commented:
@dachary Thanks for the update. i believe with this,https://github.com/ceph/ceph-deploy/pull/430 will be needed for ceph-deploy with work. @vasukulkarni

leseb commented:
@dachary No worries :)

GregMeno commented:
@leseb @dachary  Would you please help me understand the ceph-disk suite failure here?
cheers,
G

dachary commented:
@GregMeno ceph-disk activate does not have a --cluster argument but the function implementing it expects it. It does not show in the ceph-disk suite output because the part that fails is run via either systemd or udev.
